# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/arm/CodeGenerator-arm.cpp
# Commit: f0f44cfd6a07
# Full Hash: f0f44cfd6a07c787846c8c84388f503adce349aa
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-02 04:10:47
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.1: Untie callWithAbi from Frame::tls for Baseline. r=lth
#   
#   We have a tls dependency in callWithABI for wasm because we call builtins via BuiltinThunk.
#   Last one requires that tls to be loaded. In this patch I gradually extend callWithABI interface to
#   pass offset to tls. This allows us to preserve tls and load it by offset in callWithABI and frees us from
# ==============================================================================

diff -r 0e268b8d01fe -r f0f44cfd6a07 js/src/jit/arm/CodeGenerator-arm.cpp
--- a/js/src/jit/arm/CodeGenerator-arm.cpp	Thu Oct 01 04:54:42 2020 +0000
+++ b/js/src/jit/arm/CodeGenerator-arm.cpp	Thu Oct 01 07:52:07 2020 +0000
@@ -589,7 +589,7 @@
     masm.passABIArg(lhs);
     masm.passABIArg(rhs);
     masm.callWithABI(mir->bytecodeOffset(),
-                     wasm::SymbolicAddress::aeabi_idivmod);
+                     wasm::SymbolicAddress::aeabi_idivmod, mozilla::Nothing());
   } else {
     masm.setupAlignedABICall();
     masm.passABIArg(lhs);
@@ -778,7 +778,7 @@
     masm.passABIArg(lhs);
     masm.passABIArg(rhs);
     masm.callWithABI(mir->bytecodeOffset(),
-                     wasm::SymbolicAddress::aeabi_idivmod);
+                     wasm::SymbolicAddress::aeabi_idivmod, mozilla::Nothing());
   } else {
     masm.setupAlignedABICall();
     masm.passABIArg(lhs);
@@ -2249,7 +2249,8 @@
     masm.passABIArg(rhs);
     wasm::BytecodeOffset bytecodeOffset =
         (div ? div->bytecodeOffset() : mod->bytecodeOffset());
-    masm.callWithABI(bytecodeOffset, wasm::SymbolicAddress::aeabi_uidivmod);
+    masm.callWithABI(bytecodeOffset, wasm::SymbolicAddress::aeabi_uidivmod,
+                     mozilla::Nothing());
   } else {
     masm.setupAlignedABICall();
     masm.passABIArg(lhs);
@@ -2366,18 +2367,22 @@
   if (lir->mir()->isSaturating()) {
     if (lir->mir()->isUnsigned()) {
       masm.callWithABI(mir->bytecodeOffset(),
-                       wasm::SymbolicAddress::SaturatingTruncateDoubleToUint64);
+                       wasm::SymbolicAddress::SaturatingTruncateDoubleToUint64,
+                       mozilla::Nothing());
     } else {
       masm.callWithABI(mir->bytecodeOffset(),
-                       wasm::SymbolicAddress::SaturatingTruncateDoubleToInt64);
+                       wasm::SymbolicAddress::SaturatingTruncateDoubleToInt64,
+                       mozilla::Nothing());
     }
   } else {
     if (lir->mir()->isUnsigned()) {
       masm.callWithABI(mir->bytecodeOffset(),
-                       wasm::SymbolicAddress::TruncateDoubleToUint64);
+                       wasm::SymbolicAddress::TruncateDoubleToUint64,
+                       mozilla::Nothing());
     } else {
       masm.callWithABI(mir->bytecodeOffset(),
-                       wasm::SymbolicAddress::TruncateDoubleToInt64);
+                       wasm::SymbolicAddress::TruncateDoubleToInt64,
+                       mozilla::Nothing());
     }
   }
 
@@ -2433,7 +2438,7 @@
 
   MoveOp::Type result =
       toType == MIRType::Float32 ? MoveOp::FLOAT32 : MoveOp::DOUBLE;
-  masm.callWithABI(mir->bytecodeOffset(), callee, result);
+  masm.callWithABI(mir->bytecodeOffset(), callee, mozilla::Nothing(), result);
 
   DebugOnly<FloatRegister> output(ToFloatRegister(lir->output()));
   MOZ_ASSERT_IF(toType == MIRType::Double, output.value == ReturnDoubleReg);
@@ -2595,9 +2600,11 @@
   masm.passABIArg(rhs.low);
 
   if (mir->isMod()) {
-    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::ModI64);
+    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::ModI64,
+                     mozilla::Nothing());
   } else {
-    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::DivI64);
+    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::DivI64,
+                     mozilla::Nothing());
   }
 
   MOZ_ASSERT(ReturnReg64 == output);
@@ -2629,9 +2636,11 @@
   MOZ_ASSERT(gen->compilingWasm());
   MDefinition* mir = lir->mir();
   if (mir->isMod()) {
-    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::UModI64);
+    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::UModI64,
+                     mozilla::Nothing());
   } else {
-    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::UDivI64);
+    masm.callWithABI(lir->bytecodeOffset(), wasm::SymbolicAddress::UDivI64,
+                     mozilla::Nothing());
   }
 }
 